### PR TITLE
[SAI-PTF] Add ecmp test

### PR DESCRIPTION
**Pull request by zhiyuan0112**
## Description of PR
1. Config route and neighbor for nexthop group.
2. Add test_ecmp_lags.
3. Modify Ecmp to NexthopGroup.

## Test
Run all test on pipeline and they all passed.

Signed-off-by: zhiyuanliang-ms [t-zhliang@microsoft.com](mailto:t-zhliang@microsoft.com)

### CODE REVIEW

File: test/sai_test/config/route_configer.py [hunk 1]
```diff
@@ -111,13 +112,46 @@ def t0_route_config_helper(
...
+            nexthopv6=nhv6)
+        # set expected dest server
+        for item in test_obj.servers[13]:
+            item.l3_lag_obj = test_obj.dut.lag_list[2]
+            item.l3_lag_obj.neighbor_mac = test_obj.t1_list[3][100].mac
+        # set expected dest T1
+        test_obj.t1_list[3][100].l3_lag_obj = test_obj.dut.lag_list[2]
+
+        # create nexthop
+        route_configer.create_nexthops_by_lags(
```
Comment (Gfrom2016) [inline test/sai_test/config/route_configer.py:144]: This route config for ecmp is not consistent with the configuration.

File: test/sai_test/config/route_configer.py [hunk 2]
```diff
@@ -426,6 +503,61 @@ def create_nexthop_by_rif(self, rif, nexthop_device: Device):
...
+            self.client,
+            next_hop_group_id=nhop_groupv6_id,
+            next_hop_id=lag_list[0].nexthopv4_list[0].oid)
+
+        next_hop_groupv4: NexthopGroup = NexthopGroup(nhop_groupv4_id, nh_groupv4_member, [19, 20, 21, 22])
+        next_hop_groupv6: NexthopGroup = NexthopGroup(nhop_groupv6_id, nh_groupv6_member, [19, 20, 21, 22])
+
+        self.test_obj.dut.nhop_groupv4_list.append(next_hop_groupv4)
+        self.test_obj.dut.nhop_groupv6_list.append(next_hop_groupv6)
+        self.test_obj.assertEqual(status, SAI_STATUS_SUCCESS)
```
Comment (Gfrom2016) [inline test/sai_test/config/route_configer.py:557]: Where does this status variable come from?

File: test/sai_test/config/route_configer.py [hunk 3]
```diff
@@ -426,6 +503,61 @@ def create_nexthop_by_rif(self, rif, nexthop_device: Device):
...
+
+        return nexthop group for v4 and nexthop group for v6 
+        """
+        nhop_groupv4_id = sai_thrift_create_next_hop_group(self.client, type=SAI_NEXT_HOP_GROUP_TYPE_ECMP)
+        nhop_groupv6_id = sai_thrift_create_next_hop_group(self.client, type=SAI_NEXT_HOP_GROUP_TYPE_ECMP)
+
+        nh_groupv4_member = sai_thrift_create_next_hop_group_member(
+            self.client,
+            next_hop_group_id=nhop_groupv4_id,
+            next_hop_id=lag_list[0].nexthopv4_list[0].oid)
```
Comment (Gfrom2016) [inline test/sai_test/config/route_configer.py:546]: Why only use lag_list[0]?

**Comment by zhiyuan0112** [comment]
## Description of Modification
1. Group Lag1-4 following test plan
2. Add `nexthop_group` as a new attr for `Lag` Class
3. Rewrite ecmp test cases (Verify loadbalance)
4. Add an on/off switch for testing ecmp cases

## Test
Run all test on pipeline and they all passed.

Signed-off-by: zhiyuanliang-ms [t-zhliang@microsoft.com](mailto:t-zhliang@microsoft.com)

File: test/sai_test/config/route_configer.py [hunk 4]
```diff
@@ -426,6 +537,42 @@ def create_nexthop_by_rif(self, rif, nexthop_device: Device):
...
+        Create nexthop group by nexthops. Each element in all lists corresponds one by one.
+
+        return nexthop group for v4 and nexthop group for v6 
+        """
+        nhop_groupv4_id = sai_thrift_create_next_hop_group(self.client, type=SAI_NEXT_HOP_GROUP_TYPE_ECMP)
+        nhop_groupv6_id = sai_thrift_create_next_hop_group(self.client, type=SAI_NEXT_HOP_GROUP_TYPE_ECMP)
+        self.test_obj.assertEqual(self.test_obj.status(), SAI_STATUS_SUCCESS)
+
+        for nexthopv4, nexthopv6 in zip(nexthopv4_list, nexthopv6_list):
+            sai_thrift_create_next_hop_group_member(
```
Comment (Gfrom2016) [inline test/sai_test/config/route_configer.py:551]: Check status after this function

File: test/sai_test/config/route_configer.py [hunk 5]
```diff
@@ -426,6 +537,42 @@ def create_nexthop_by_rif(self, rif, nexthop_device: Device):
...
+                next_hop_group_id=nhop_groupv4_id,
+                next_hop_id=nexthopv4.oid)
+            sai_thrift_create_next_hop_group_member(
+                self.client,
+                next_hop_group_id=nhop_groupv6_id,
+                next_hop_id=nexthopv6.oid)
+            self.test_obj.assertEqual(self.test_obj.status(), SAI_STATUS_SUCCESS)
+
+        member_port_indexs = []
+        for lag in lag_list:
```
Comment (Gfrom2016) [inline test/sai_test/config/route_configer.py:562]: set lag.nexthop_group somewhere

File: test/sai_test/sai_ecmp_test.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        """
+        T0TestBase.setUp(self, is_create_route_for_nhopgrp=True)
+        
+    def test_nhopgrp_configv4(self):
+        """
+        1. Get those attributes [number_of_ecmp_groups, ecmp_members, available_next_hop_group_entry, available_next_hop_group_member_entry]
+        2. remove ecmp member and next group member
+        3. Get attributes again and check the value
+        """
+        print("Lag l3 load balancing test based on src IP")
```
Comment (Gfrom2016) [inline test/sai_test/sai_ecmp_test.py:22]: printed msg do not match the case

File: test/sai_test/config/route_configer.py [hunk 7]
```diff
@@ -426,6 +537,42 @@ def create_nexthop_by_rif(self, rif, nexthop_device: Device):
 
         return nhopv4, nhopv6
 
+    def create_nexthop_group_by_nexthops(self, nexthopv4_list: List[Nexthop], nexthopv6_list: List[Nexthop], lag_list: List[Lag], dest_device: Device):
+        """
+        Create nexthop group by nexthops. Each element in all lists corresponds one by one.
+
+        return nexthop group for v4 and nexthop group for v6 
+        """
+        nhop_groupv4_id = sai_thrift_create_next_hop_group(self.client, type=SAI_NEXT_HOP_GROUP_TYPE_ECMP)
```
Comment (Gfrom2016) [inline test/sai_test/config/route_configer.py:546]: Check status after this.

File: test/sai_test/sai_ecmp_test.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        max_itrs = 150
+        begin_port = 2000
+        member_port_indexs = self.t1_list[1][100].l3_lag_obj.member_port_indexs +\
```
Comment (Gfrom2016) [inline test/sai_test/sai_ecmp_test.py:25]: NexthopGroup already has a member_port_indexs variable

File: test/sai_test/sai_ecmp_test.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
+from sai_test_base import T0TestBase
```
Comment (Gfrom2016) [inline test/sai_test/sai_ecmp_test.py:1]: format code for this file

**Pull request closed without merging**